{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nobackup1c/users/payette/eggNOGbyClass3\n"
     ]
    }
   ],
   "source": [
    "import ete3\n",
    "import re\n",
    "import itertools\n",
    "import multiprocessing\n",
    "import random\n",
    "\n",
    "import pandas as pd\n",
    "import numpy  as np\n",
    "import igraph as ig\n",
    "import pickle as pkl\n",
    "\n",
    "from scipy.spatial.distance import squareform, pdist\n",
    "from scipy.stats            import mannwhitneyu\n",
    "from collections            import Counter\n",
    "\n",
    "ncbi = ete3.NCBITaxa()\n",
    "%cd ~/work2/eggNOGbyClass3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_genomes = pd.read_csv('genomes.tab',\n",
    "                              sep='\\t',\n",
    "                              index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ABCDG_sampled_genomes = pd.read_excel('ABGD_tree_taxa.xlsx')  #New taxa file v.4 with 31 new rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ABCDG_sampled_genomes = ABCDG_sampled_genomes[ABCDG_sampled_genomes['TaxID'] != 2762020] #Exclude this taxa b/c error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1097.0, 1219.0, 33072.0, 198252.0]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(pd.DataFrame(sampled_genomes.species_taxid).species_taxid.head().unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "lineages = pd.DataFrame()\n",
    "for taxid in ABCDG_sampled_genomes['TaxID'].unique():\n",
    "    if pd.isna(taxid):\n",
    "        continue\n",
    "    lineages = lineages.append({tax_rank: tmp_taxid \n",
    "                                 for tmp_taxid, tax_rank in ncbi.get_rank(ncbi.get_lineage(taxid)).items()},\n",
    "                                ignore_index=True)\n",
    "lineages = lineages.reindex(columns=['class', 'family',  'genus', 'phylum',\n",
    "                                     'order', 'species', 'superkingdom']).copy()\n",
    "lineages = lineages.query('superkingdom == 2').copy()\n",
    "\n",
    "lineages.loc[lineages['phylum']==1224, 'phylum'] = lineages.loc[lineages['phylum']==1224, 'class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>species</th>\n",
       "      <th>superkingdom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>191410.0</td>\n",
       "      <td>191412.0</td>\n",
       "      <td>256319.0</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>191411.0</td>\n",
       "      <td>1097.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>191410.0</td>\n",
       "      <td>191412.0</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>191411.0</td>\n",
       "      <td>1092.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>191410.0</td>\n",
       "      <td>191412.0</td>\n",
       "      <td>1101.0</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>191411.0</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>191410.0</td>\n",
       "      <td>191412.0</td>\n",
       "      <td>100715.0</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>191411.0</td>\n",
       "      <td>100716.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>191410.0</td>\n",
       "      <td>191412.0</td>\n",
       "      <td>1099.0</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>191411.0</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>191410.0</td>\n",
       "      <td>191412.0</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>191411.0</td>\n",
       "      <td>337090.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        class    family     genus  phylum     order   species  superkingdom\n",
       "101  191410.0  191412.0  256319.0  1090.0  191411.0    1097.0           2.0\n",
       "102  191410.0  191412.0    1091.0  1090.0  191411.0    1092.0           2.0\n",
       "103  191410.0  191412.0    1101.0  1090.0  191411.0    1102.0           2.0\n",
       "104  191410.0  191412.0  100715.0  1090.0  191411.0  100716.0           2.0\n",
       "204  191410.0  191412.0    1099.0  1090.0  191411.0    1100.0           2.0\n",
       "221  191410.0  191412.0    1091.0  1090.0  191411.0  337090.0           2.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lineages.loc[lineages['phylum']==1090] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_groups  = pd.read_parquet('working_eggNOG_groups.parquet')\n",
    "working_trees   = pd.read_parquet('working_eggNOG_trees.parquet' )\n",
    "eggNOG_taxonomy = pd.read_parquet('eggNOG_taxonomy.parquet'      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pairwise_distances(group_id):\n",
    "    \n",
    "    tree = ete3.Tree(working_trees.loc[group_id, 'tree'])\n",
    "\n",
    "    leaf_names = []\n",
    "    for count, node in enumerate(tree.traverse()):\n",
    "        if node.is_leaf():\n",
    "            leaf_names.append(node.name)\n",
    "        else:\n",
    "            node.name = 'node_%i' % count\n",
    "    leaf_names = np.array(leaf_names)\n",
    "\n",
    "    nodes         = []\n",
    "    children      = []\n",
    "    branch_length = []\n",
    "    for node in tree.traverse():\n",
    "        if not node.is_leaf():\n",
    "            for child in node.get_children():\n",
    "                nodes.append(         node.name)\n",
    "                children.append(     child.name)\n",
    "                branch_length.append(child.dist)\n",
    "\n",
    "    branch_length_df                  = pd.DataFrame()\n",
    "    branch_length_df['node']          = nodes\n",
    "    branch_length_df['child']         = children\n",
    "    branch_length_df['branch_length'] = branch_length\n",
    "\n",
    "    dag  = ig.Graph.TupleList(edges=branch_length_df[['node', \n",
    "                                                      'child', \n",
    "                                                      'branch_length']].itertuples(index=False), \n",
    "                                directed=False, \n",
    "                                weights=True)\n",
    "    \n",
    "    dist_matrix = pd.DataFrame(index  =leaf_names, \n",
    "                               columns=leaf_names, \n",
    "                               data   =np.array(dag.shortest_paths(source=leaf_names, \n",
    "                                                                   target=leaf_names, \n",
    "                                                                   weights='weight'))\n",
    "                              )\n",
    "    return(dist_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_taxa_graph(dist_matrix, phyla):\n",
    "    triu_indices       = np.triu_indices_from(dist_matrix, k=1)\n",
    "    \n",
    "    edge_list                 = pd.DataFrame()\n",
    "    edge_list['phylum1']      = phyla[triu_indices[0]]\n",
    "    edge_list['phylum2']      = phyla[triu_indices[1]]\n",
    "    edge_list['sequence1']    = dist_matrix.index[triu_indices[0]]\n",
    "    edge_list['sequence2']    = dist_matrix.index[triu_indices[1]]\n",
    "    edge_list['distance']     = dist_matrix.values[triu_indices]\n",
    "    edge_list['inverse_dist'] = np.e**np.negative(edge_list.distance)\n",
    "\n",
    "    graph  = ig.Graph.TupleList(edges=edge_list[['sequence1', \n",
    "                                                 'sequence2', \n",
    "                                                 'inverse_dist']].itertuples(index=False), \n",
    "                                directed=False, \n",
    "                                weights =True)\n",
    "    \n",
    "    return(edge_list, graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cles(lessers, greaters):\n",
    "    #\n",
    "    # https://github.com/ajschumacher/cles/blob/master/cles.py\n",
    "    #\n",
    "    \"\"\"Common-Language Effect Size\n",
    "    Probability that a random draw from `greater` is in fact greater\n",
    "    than a random draw from `lesser`.\n",
    "    Args:\n",
    "      lesser, greater: Iterables of comparables.\n",
    "    \"\"\"\n",
    "    if len(lessers) == 0 and len(greaters) == 0:\n",
    "        raise ValueError('At least one argument must be non-empty')\n",
    "    # These values are a bit arbitrary, but make some sense.\n",
    "    # (It might be appropriate to warn for these cases.)\n",
    "    if len(lessers) == 0:\n",
    "        return 1\n",
    "    if len(greaters) == 0:\n",
    "        return 0\n",
    "    numerator = 0\n",
    "    lessers, greaters = sorted(lessers), sorted(greaters)\n",
    "    lesser_index = 0\n",
    "    for greater in greaters:\n",
    "        while lesser_index < len(lessers) and lessers[lesser_index] < greater:\n",
    "            lesser_index += 1\n",
    "        numerator += lesser_index  # the count less than the greater\n",
    "    denominator = len(lessers) * len(greaters)\n",
    "    return float(numerator) / denominator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assess_cluster(reference_phylum, minimal_freq_phyla, cluster_edges, cluster_nodes):\n",
    "\n",
    "    #\n",
    "    # store distances between reference phylum and others\n",
    "    cluster_dists = pd.DataFrame(columns=['phylum', 'median', 'distances'])\n",
    "\n",
    "    #\n",
    "    # traverse all phylum pairs containing the reference phylum\n",
    "    for phylum in minimal_freq_phyla:\n",
    "        if phylum == reference_phylum:\n",
    "            continue\n",
    "\n",
    "        inter_phyla = cluster_edges.loc[((cluster_edges.phylum1==reference_phylum) & (cluster_edges.phylum2==phylum)) |\\\n",
    "                                        ((cluster_edges.phylum2==reference_phylum) & (cluster_edges.phylum1==phylum))]\n",
    "        \n",
    "        #\n",
    "        # create a quadratic matrix of pairwise distances between phyla\n",
    "        #   rows and columns must be unique pairs of sequence names\n",
    "        indices     = np.unique(inter_phyla[['sequence1', 'sequence2']])\n",
    "        adjacencies = pd.DataFrame(index  =indices, \n",
    "                                   columns=indices,\n",
    "                                   data   =0.0)\n",
    "        \n",
    "        #\n",
    "        # add distances between sequences from the edge list to the quadratic matrix\n",
    "        #   as the matrix is quadratic, add values to both directions\n",
    "        indexer     = adjacencies.index.get_indexer\n",
    "        adjacencies.values[indexer(inter_phyla.sequence1), indexer(inter_phyla.sequence2)] = inter_phyla.distance.values\n",
    "        adjacencies.values[indexer(inter_phyla.sequence2), indexer(inter_phyla.sequence1)] = inter_phyla.distance.values\n",
    "\n",
    "        #\n",
    "        # sum the obtained distances into a single cell\n",
    "        tmp_closest_to_phylum = adjacencies.loc[cluster_nodes.loc[cluster_nodes.phylum==reference_phylum, 'name'],\n",
    "                                                cluster_nodes.loc[cluster_nodes.phylum==phylum,           'name']].sum()\n",
    "        tmp_closest_to_phylum.sort_values(inplace=True)\n",
    "        #\n",
    "        # and grab the five sequences from <phylum> closest to all from <reference phylum>\n",
    "        tmp_closest_to_phylum = tmp_closest_to_phylum.index[:5]\n",
    "\n",
    "        try:\n",
    "            #\n",
    "            # get all inter-phyla distances between\n",
    "            distances_to_reference_phylum = adjacencies.loc[\n",
    "                #\n",
    "                # all sequences from <reference phylum>\n",
    "                cluster_nodes.loc[cluster_nodes.phylum==reference_phylum, 'name'],\n",
    "                #\n",
    "                # the 5 sequences from <phylum> closest to <reference phylum>\n",
    "                tmp_closest_to_phylum\n",
    "            ].values.flatten()\n",
    "        except IndexError:\n",
    "            continue        \n",
    "\n",
    "        cluster_dists = cluster_dists.append(pd.Series(data =[phylum, \n",
    "                                                              np.median(distances_to_reference_phylum), \n",
    "                                                              distances_to_reference_phylum], \n",
    "                                                       index=['phylum', 'median', 'distances']),\n",
    "                                             ignore_index=True)\n",
    "    return(cluster_dists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_phyla_evol_distances(group_id):    \n",
    "    dist_matrix = get_pairwise_distances(group_id)\n",
    "\n",
    "    taxids = [int(leaf.split('.')[0]) for leaf in dist_matrix.index]\n",
    "    phyla  = eggNOG_taxonomy.loc[taxids, 'phylum'].values.astype(int)\n",
    "\n",
    "    edge_list, graph  = create_taxa_graph(dist_matrix, phyla)\n",
    "\n",
    "    random.seed(12345)\n",
    "    clusters = graph.community_multilevel(weights='weight')\n",
    "\n",
    "    node_data = pd.DataFrame(columns=['name', 'phylum', 'cluster'],\n",
    "                             data   =zip(dist_matrix.index, \n",
    "                                         phyla, \n",
    "                                         clusters.membership)\n",
    "                            )\n",
    "    \n",
    "    cluster_evol_relations = {}\n",
    "    \n",
    "    target_phyla = {1090,    # chlorobi\n",
    "                    1117,    # cyanobacteria\n",
    "                    200795,  # chloroflexi\n",
    "                    976,     # bacteroidetes #\n",
    "                    1134404, # ignavibacteriae #\n",
    "                    1798710, # melainabacteria #\n",
    "                    28211,   # Alphaproteobacteria\n",
    "                    28216,   # Betaproteobacteria\n",
    "                    1236,    # Gammaproteobacteria\n",
    "                    28221,   # Deltaproteobacteria\n",
    "                    29547,   # Epsilonproteobacteria #\n",
    "                    580370,  # Zetaproteobacteria #\n",
    "                    1807140} # Acidithiobacillia #\n",
    "    \n",
    "    #target_phyla = { 28211,          # Alpha\n",
    "    #                 28216,          # Beta\n",
    "    #                 28221,          # Delta\n",
    "    #                 1236            # Gamma\n",
    "    #               }\n",
    "    \n",
    "    #target_phyla = {1090,    # chlorobi\n",
    "    #                1117,    # cyanobacteria\n",
    "    #                1224,    # proteobacteria\n",
    "    #                200795,  # chloroflexi\n",
    "    #                976,     # bacteroidetes\n",
    "    #                1134404, # ignavibacteriae\n",
    "    #                1798710} # melainabacteria\n",
    "\n",
    "    for cluster_num in set(clusters.membership):\n",
    "        \n",
    "        cluster_nodes      = node_data[node_data.cluster==cluster_num]\n",
    "        minimal_freq_phyla = [phylum \n",
    "                              for phylum, frequency in Counter(cluster_nodes.phylum).items() \n",
    "                              if frequency>=5   # at least five sequences from a phylum\n",
    "                              and phylum > 0]   # given pandas manipulation, unknown phyla are represented \n",
    "                                                #   as NAN, which when forced to be int are negative numbers\n",
    "                                                #   that is why we ignore phylum taxids smaller than zero...\n",
    "       \n",
    "        #\n",
    "        # if there are fewer than two phyla of interested within the tree cluster there is no reason to\n",
    "        #   assess it...\n",
    "        if len( target_phyla.intersection( minimal_freq_phyla ) ) < 2:\n",
    "            continue\n",
    "        \n",
    "        cluster_evol_relations[cluster_num] = {}\n",
    "        \n",
    "        #\n",
    "        # filter patristic distances from the whole tree to only those between sequences within the cluster\n",
    "        cluster_edges = edge_list.loc[(edge_list.sequence1.isin(cluster_nodes.name)) &\n",
    "                                      (edge_list.sequence2.isin(cluster_nodes.name)),\n",
    "                                      ['phylum1', 'phylum2', 'sequence1', 'sequence2', 'distance']]\n",
    "\n",
    "        #\n",
    "        # filter again, leaving only between sequences whose phylum fulfills the minimal frequency\n",
    "        cluster_edges      = cluster_edges[(cluster_edges.phylum1.isin(minimal_freq_phyla)) &\\\n",
    "                                           (cluster_edges.phylum2.isin(minimal_freq_phyla))]\n",
    "        #\n",
    "        # we will divide all pairwise distances by the cluster's mean\n",
    "        normalizer         = np.median(cluster_edges.distance)\n",
    "        #\n",
    "        # remove all intra-phylum distances\n",
    "        cluster_edges      = cluster_edges[cluster_edges.phylum1 != cluster_edges.phylum2] \n",
    "\n",
    "        #\n",
    "        # assess all inter-phyla distances, using each phylum as a reference to itself\n",
    "        #\n",
    "        for ref_phylum in target_phyla.intersection(minimal_freq_phyla):\n",
    "            cluster_dists = assess_cluster(ref_phylum, \n",
    "                                           minimal_freq_phyla, \n",
    "                                           cluster_edges,\n",
    "                                           cluster_nodes)\n",
    "\n",
    "            #\n",
    "            # sort phyla by its avg. distance to <reference phylum>\n",
    "            cluster_dists.sort_values('median', inplace=True)\n",
    "            cluster_evol_relations[cluster_num][ref_phylum] = {\n",
    "                'df'         :cluster_dists[['phylum', 'median']].copy(),\n",
    "                'significant':False\n",
    "            }\n",
    "            if not cluster_dists.shape[0]:\n",
    "                continue\n",
    "\n",
    "            cluster_evol_relations[cluster_num][ref_phylum]['df']['median']   /= normalizer\n",
    "            \n",
    "            #\n",
    "            # if there is only one <phylum> together with <reference phylum>, the proximity between\n",
    "            #   both is automaticaly significant\n",
    "            if cluster_dists.shape[0] == 1:\n",
    "                cluster_evol_relations[cluster_num][ref_phylum]['significant'] = True\n",
    "                continue\n",
    "\n",
    "            try:\n",
    "                #\n",
    "                # test if distances from the closest phylum to <reference phylum> is significantly\n",
    "                #   smaller than distances from the second closest phylum\n",
    "                hypothesis = mannwhitneyu(cluster_dists.iloc[0, 2], \n",
    "                                          cluster_dists.iloc[1, 2], \n",
    "                                          alternative='less')\n",
    "            except ValueError:\n",
    "                continue\n",
    "            else:\n",
    "                #\n",
    "                # both cles lines below should work identically... I am leaving the above because it is the one I\n",
    "                #   used in the paper, no real reason...\n",
    "                effect_size = hypothesis.statistic / (len(cluster_dists.iloc[0, 2])*len(cluster_dists.iloc[1, 2]))\n",
    "#                 effect_size = 1-cles(cluster_dists.iloc[0, 2], cluster_dists.iloc[1, 2])\n",
    "\n",
    "                if hypothesis.pvalue < 0.01 and effect_size < 0.2:\n",
    "                    cluster_evol_relations[cluster_num][ref_phylum]['significant'] = True\n",
    "    \n",
    "    return(group_id, cluster_evol_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26832\n"
     ]
    }
   ],
   "source": [
    "print(len(working_groups.group_id.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 52.9 s, sys: 13.1 s, total: 1min 6s\n",
      "Wall time: 12h 47min 41s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "pool    = multiprocessing.Pool(processes=10, maxtasksperchild=5)\n",
    "results = pool.map_async(get_phyla_evol_distances, working_groups.group_id.values)\n",
    "pool.close()\n",
    "pool.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('ABGD_new_phyla_results.pkl', 'wb') as out:\n",
    "    pkl.dump(results.get(), out)\n",
    "del(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('COG0480',\n",
       " {0: {1090: {'df':      phylum    median\n",
       "    14      976  0.501242\n",
       "    0      1239  0.519405\n",
       "    6    508458  0.536891\n",
       "    10    74201  0.544430\n",
       "    8     28221  0.566401\n",
       "    4    200795  0.576228\n",
       "    1    200918  0.608968\n",
       "    15   200930  0.611884\n",
       "    2    201174  0.642499\n",
       "    9     57723  0.648633\n",
       "    19    32066  0.660574\n",
       "    7    203691  0.662202\n",
       "    12     1297  0.676243\n",
       "    3      1117  0.683408\n",
       "    11   203682  0.711719\n",
       "    5      1236  0.739330\n",
       "    16  1553900  0.739450\n",
       "    20   544448  0.741396\n",
       "    17   200940  0.762089\n",
       "    18    28216  0.775751\n",
       "    13    28211  0.913984,\n",
       "    'significant': False},\n",
       "   976: {'df':      phylum    median\n",
       "    14     1090  0.635604\n",
       "    0      1239  0.690213\n",
       "    10    74201  0.697829\n",
       "    6    508458  0.706834\n",
       "    8     28221  0.727624\n",
       "    4    200795  0.752764\n",
       "    1    200918  0.777809\n",
       "    15   200930  0.781151\n",
       "    2    201174  0.812368\n",
       "    9     57723  0.826613\n",
       "    19    32066  0.833119\n",
       "    7    203691  0.833278\n",
       "    12     1297  0.850344\n",
       "    3      1117  0.853479\n",
       "    11   203682  0.880876\n",
       "    5      1236  0.895748\n",
       "    17   200940  0.899000\n",
       "    16  1553900  0.909816\n",
       "    20   544448  0.912124\n",
       "    18    28216  0.946615\n",
       "    13    28211  1.034646,\n",
       "    'significant': False},\n",
       "   28211: {'df':      phylum    median\n",
       "    3      1117  0.591735\n",
       "    2    201174  0.801770\n",
       "    0      1239  0.879160\n",
       "    4    200795  0.907449\n",
       "    6    508458  0.919743\n",
       "    8     28221  0.921314\n",
       "    12     1297  0.937108\n",
       "    11   203682  0.949479\n",
       "    9     57723  0.959198\n",
       "    1    200918  0.966341\n",
       "    5      1236  0.968972\n",
       "    15   200930  0.969261\n",
       "    18    28216  0.983144\n",
       "    19    32066  1.016384\n",
       "    7    203691  1.022450\n",
       "    14     1090  1.039333\n",
       "    13      976  1.085405\n",
       "    10    74201  1.086168\n",
       "    16  1553900  1.097232\n",
       "    20   544448  1.101102\n",
       "    17   200940  1.119428,\n",
       "    'significant': True},\n",
       "   1236: {'df':      phylum    median\n",
       "    18    28216  0.324316\n",
       "    10   203682  0.561012\n",
       "    0      1239  0.639756\n",
       "    4    200795  0.683385\n",
       "    2    201174  0.683610\n",
       "    5    508458  0.685150\n",
       "    7     28221  0.685770\n",
       "    11     1297  0.709970\n",
       "    15   200930  0.727185\n",
       "    1    200918  0.729227\n",
       "    6    203691  0.782729\n",
       "    19    32066  0.785847\n",
       "    8     57723  0.793241\n",
       "    3      1117  0.793589\n",
       "    14     1090  0.803805\n",
       "    9     74201  0.838781\n",
       "    13      976  0.844147\n",
       "    16  1553900  0.851644\n",
       "    20   544448  0.861498\n",
       "    17   200940  0.876002\n",
       "    12    28211  0.892820,\n",
       "    'significant': False},\n",
       "   1117: {'df':      phylum    median\n",
       "    11     1297  0.344222\n",
       "    7     28221  0.358766\n",
       "    16  1553900  0.459369\n",
       "    10   203682  0.472203\n",
       "    13      976  0.509036\n",
       "    6    203691  0.517163\n",
       "    0      1239  0.536996\n",
       "    4      1236  0.572331\n",
       "    9     74201  0.589131\n",
       "    15   200930  0.603816\n",
       "    5    508458  0.615346\n",
       "    3    200795  0.634128\n",
       "    1    200918  0.648976\n",
       "    8     57723  0.666939\n",
       "    19    32066  0.675293\n",
       "    2    201174  0.691477\n",
       "    17   200940  0.714960\n",
       "    14     1090  0.736063\n",
       "    20   544448  0.758677\n",
       "    18    28216  0.831315\n",
       "    12    28211  0.944158,\n",
       "    'significant': False},\n",
       "   28216: {'df':      phylum    median\n",
       "    5      1236  0.287816\n",
       "    11   203682  0.573309\n",
       "    0      1239  0.662874\n",
       "    4    200795  0.692251\n",
       "    2    201174  0.696527\n",
       "    6    508458  0.701967\n",
       "    8     28221  0.710703\n",
       "    12     1297  0.721582\n",
       "    1    200918  0.749124\n",
       "    16   200930  0.753121\n",
       "    9     57723  0.796021\n",
       "    19    32066  0.801169\n",
       "    7    203691  0.804148\n",
       "    15     1090  0.823068\n",
       "    3      1117  0.824792\n",
       "    14      976  0.863446\n",
       "    10    74201  0.865961\n",
       "    17  1553900  0.880693\n",
       "    20   544448  0.885089\n",
       "    18   200940  0.899011\n",
       "    13    28211  0.907601,\n",
       "    'significant': True},\n",
       "   200795: {'df':      phylum    median\n",
       "    0      1239  0.444074\n",
       "    5    508458  0.482618\n",
       "    7     28221  0.487299\n",
       "    1    200918  0.528619\n",
       "    15   200930  0.529696\n",
       "    2    201174  0.531559\n",
       "    8     57723  0.576863\n",
       "    6    203691  0.581559\n",
       "    19    32066  0.586421\n",
       "    11     1297  0.587198\n",
       "    3      1117  0.600963\n",
       "    14     1090  0.603141\n",
       "    10   203682  0.628198\n",
       "    4      1236  0.635619\n",
       "    13      976  0.647378\n",
       "    9     74201  0.647412\n",
       "    16  1553900  0.659101\n",
       "    20   544448  0.665126\n",
       "    18    28216  0.667308\n",
       "    17   200940  0.685835\n",
       "    12    28211  0.805500,\n",
       "    'significant': False},\n",
       "   28221: {'df':      phylum    median\n",
       "    11     1297  0.543243\n",
       "    3      1117  0.545842\n",
       "    0      1239  0.566380\n",
       "    10   203682  0.591561\n",
       "    16  1553900  0.594032\n",
       "    15   200930  0.623318\n",
       "    13      976  0.626484\n",
       "    17   200940  0.642860\n",
       "    6    508458  0.643346\n",
       "    4    200795  0.660790\n",
       "    9     74201  0.661550\n",
       "    5      1236  0.663832\n",
       "    8     57723  0.665379\n",
       "    7    203691  0.671608\n",
       "    2    201174  0.672339\n",
       "    1    200918  0.676129\n",
       "    19    32066  0.702237\n",
       "    14     1090  0.764618\n",
       "    20   544448  0.788035\n",
       "    18    28216  0.859777\n",
       "    12    28211  0.914040,\n",
       "    'significant': False}},\n",
       "  3: {1090: {'df':     phylum    median\n",
       "    2     1239  0.727284\n",
       "    5   200783  0.794387\n",
       "    4    28221  0.796660\n",
       "    9   201174  0.866679\n",
       "    7   200940  0.873344\n",
       "    8   204428  0.893253\n",
       "    10  200795  0.918461\n",
       "    1   508458  0.960985\n",
       "    6   203691  0.964614\n",
       "    0   200918  1.013889\n",
       "    3    74201  1.136366\n",
       "    11  544448  1.367518,\n",
       "    'significant': True},\n",
       "   200795: {'df':     phylum    median\n",
       "    2     1239  0.762876\n",
       "    4    28221  0.846173\n",
       "    5   200783  0.864812\n",
       "    10  201174  0.876514\n",
       "    7   200940  0.925763\n",
       "    8     1090  0.954879\n",
       "    6   203691  1.015562\n",
       "    1   508458  1.019312\n",
       "    9   204428  1.025407\n",
       "    0   200918  1.060484\n",
       "    3    74201  1.187682\n",
       "    11  544448  1.400600,\n",
       "    'significant': False},\n",
       "   28221: {'df':     phylum    median\n",
       "    6   200940  0.605995\n",
       "    2     1239  0.610726\n",
       "    4   200783  0.658386\n",
       "    5   203691  0.772706\n",
       "    10  200795  0.794580\n",
       "    7     1090  0.795913\n",
       "    9   201174  0.805672\n",
       "    8   204428  0.856507\n",
       "    1   508458  0.857393\n",
       "    0   200918  0.873885\n",
       "    3    74201  1.015639\n",
       "    11  544448  1.248728,\n",
       "    'significant': False}},\n",
       "  5: {1807140: {'df':    phylum    median\n",
       "    0    1236  0.606336\n",
       "    4   28216  0.661868\n",
       "    3   28221  0.759117\n",
       "    1   29547  0.802778\n",
       "    2  200930  0.851005,\n",
       "    'significant': True},\n",
       "   29547: {'df':     phylum    median\n",
       "    3    28221  0.746421\n",
       "    2   200930  0.830656\n",
       "    0     1236  0.907130\n",
       "    1  1807140  0.933739\n",
       "    4    28216  0.982194,\n",
       "    'significant': False},\n",
       "   1236: {'df':     phylum    median\n",
       "    4    28216  0.881453\n",
       "    0  1807140  0.957183\n",
       "    3    28221  1.102599\n",
       "    1    29547  1.144877\n",
       "    2   200930  1.191363,\n",
       "    'significant': False},\n",
       "   28216: {'df':     phylum    median\n",
       "    0     1236  0.638277\n",
       "    1  1807140  0.817994\n",
       "    4    28221  0.961185\n",
       "    2    29547  0.999745\n",
       "    3   200930  1.050469,\n",
       "    'significant': True},\n",
       "   28221: {'df':     phylum    median\n",
       "    3   200930  0.816820\n",
       "    2    29547  0.866007\n",
       "    0     1236  1.001703\n",
       "    1  1807140  1.011381\n",
       "    4    28216  1.060306,\n",
       "    'significant': False}},\n",
       "  6: {28216: {'df':    phylum    median\n",
       "    2   28211  0.956230\n",
       "    3    1236  1.144941\n",
       "    0  201174  1.170944\n",
       "    4    1117  1.184927\n",
       "    1   57723  1.208555,\n",
       "    'significant': True},\n",
       "   28211: {'df':    phylum    median\n",
       "    0  201174  0.915822\n",
       "    1   57723  0.939777\n",
       "    2    1236  1.050621\n",
       "    4    1117  1.072041\n",
       "    3   28216  1.193246,\n",
       "    'significant': False},\n",
       "   1236: {'df':    phylum    median\n",
       "    4    1117  0.470656\n",
       "    2   28211  1.054291\n",
       "    0  201174  1.220100\n",
       "    1   57723  1.248409\n",
       "    3   28216  1.344421,\n",
       "    'significant': True},\n",
       "   1117: {'df':    phylum    median\n",
       "    3    1236  0.317027\n",
       "    2   28211  0.936024\n",
       "    0  201174  1.102575\n",
       "    1   57723  1.131241\n",
       "    4   28216  1.228444,\n",
       "    'significant': True}}})"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_phyla_evol_distances('COG0480')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
